algoliasearch/src/main/java/com/algolia/search/saas/BaseIndex.java

```diff
@@ -330,24 +332,37 @@ protected void deleteByQuery(Query query) throws AlgoliaException {
 
     /**
      * Search inside the index
+     *
      * @return a JSONObject containing search results
      * @throws AlgoliaException
      */
-    protected JSONObject search(Query query) throws AlgoliaException {
-        String paramsString = query.getQueryString();
-        if (paramsString.length() > 0) {
-            return client.getRequest("/1/indexes/" + encodedIndexName + "?" + paramsString, true);
-        } else {
-            return client.getRequest("/1/indexes/" + encodedIndexName, true);
+    protected JSONObject search(@NonNull Query query) throws AlgoliaException {
+        String cacheKey = null;
+        byte[] rawResponse = null;
+        if (isCacheEnabled) {
+            cacheKey = String.format("%s", query.getQueryString());
+            rawResponse = searchCache.get(cacheKey);
+        }
+        try {
+            if (rawResponse == null) {
+                rawResponse = searchRaw(query);
+                if (isCacheEnabled) {
+                    searchCache.put(cacheKey, rawResponse);
+                }
+            }
+            return BaseAPIClient._getJSONObject(rawResponse);
```

Comment:
As commented on a previous version: I would rather store the response when we are sure it parses correctly.